drift-transport-netty/src/main/java/com/facebook/drift/transport/netty/codec/ZlibTransformer.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2013 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.drift.transport.netty.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+public class ZlibTransformer
+        implements ByteBufTransformer
+{
+    private static final int ZLIB_COMPRESSION_BUFFER_SIZE = 512;
+    public static final ZlibTransformer ZLIB_TRANSFORMER = new ZlibTransformer();
+
+    @Override
+    public ByteBuf transform(ByteBuf data)
+    {
+        return compress(data);
+    }
+
+    @Override
+    public ByteBuf untransform(ByteBuf data)
+    {
+        return uncompress(data);
+    }
+
+    private static ByteBuf compress(ByteBuf data)
+    {
+        Deflater compressor = new Deflater();
+        try {
+            int inputLength = data.readableBytes();
+            byte[] output = new byte[inputLength + ZLIB_COMPRESSION_BUFFER_SIZE]; // output might be larger
+            byte[] input;
+            int inputOffset;
+            if (data.hasArray()) {
+                input = data.array();
+                inputOffset = data.readerIndex();
+            }
+            else {
+                input = new byte[inputLength];
+                data.readBytes(input);
+                inputOffset = 0;
+            }
+            compressor.setInput(input, inputOffset, inputLength);
+            compressor.finish();
+            int length = compressor.deflate(output);
+            if (!compressor.finished()) {
+                // Output buffer was not big enough.  Unlikely.
+                // If you hit this, you probably shouldn't be using ZLIB_TRANSFORM :)
+                throw new IllegalArgumentException("Output compress buffer not big enough");
+            }
+            return Unpooled.wrappedBuffer(output, 0, length);
+        }
+        finally {
+            compressor.end();
+            data.release();
```

Comment:
compress/uncompress took in a ByteBuf and return a new `ByteBuf`. To allow resource correctly released, we need to manually reduce the reference count on old ByteBuf.